### PR TITLE
Forward 'X-Origin' request header

### PR DESCRIPTION
**Pull request by ivanruski**
# Description

Forward 'X-Origin' request header from incoming requests to outgoing requests to `ontotext-suggestions-api` and `authors-suggestion-api`

## Why

[JIRA](https://financialtimes.atlassian.net/browse/UPPSF-3062)

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [x] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: reqorigin/reqorigin.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Package reqorigin provides functions for extracting 'X-Origin' header from
+// http request.
+package reqorigin
```
Comment (ivan-p-nikolov) [inline reqorigin/reqorigin.go:3]: **question:** ‏What is the reasoning behind this package? 

Why not put this in the suggester?

File: service/aggregator.go [hunk 2]
```diff
@@ -99,22 +108,30 @@ func (s *AggregateSuggester) GetSuggestions(payload []byte, tid string) (Suggest
...
-			}
-		}
+	for i := range s.Suggesters {
+		filteredSuggestions := filterBlacklistedSuggestions(responseMap[i], blacklist, s.Blacklister)
+		aggregateResp.Suggestions = append(aggregateResp.Suggestions, filteredSuggestions...)
 	}
 	return aggregateResp, nil
 }
 
+func filterBlacklistedSuggestions(suggestions []Suggestion, list Blacklist, blacklister ConceptBlacklister) []Suggestion {
```
Comment (ivan-p-nikolov) [inline service/aggregator.go:118]:
```suggestion
func filterDisallowedSuggestions(suggestions []Suggestion, list Blacklist, blacklister ConceptBlacklister) []Suggestion {
```

File: service/aggregator.go [hunk 3]
```diff
@@ -28,7 +28,16 @@ func NewAggregateSuggester(log *logger.UPPLogger, concordance *ConcordanceServic
 	}
 }
 
-func (s *AggregateSuggester) GetSuggestions(payload []byte, tid string) (SuggestionsResponse, error) {
+// GetSuggestions calls several services to build it's return value.
+//
+// It calls concurrently ontotext-suggestions-api, authors-suggestion-api and
```
Comment (ivan-p-nikolov) [inline service/aggregator.go:33]: **suggestion(non-blocking):** I would reword the comment. 

It is not wrong, but it is not relevant to the method.
In the code we use `Suggesters`, `BroaderProvider` and `Blacklister`, not `ontotext-suggestions-api`, `authors-suggestion-api`, `public-things-api` and `concept-suggestions-blacklister`.

What service we hit is depended on how we setup the aggregator, not how the aggregator works.  If you understand what I mean.

**Review by ivan-p-nikolov — DISMISSED** [review]
LGTM

File: reqorigin/reqorigin.go (hunk 1 above)
Comment (ivanruski) [reply reqorigin/reqorigin.go:3]: I've got the idea for this package while reading about contexts. I thought that later I can extend this package with context related functions and use them in suggester. Just like the tidutils package.

I think that 'X-Origin' is request scoped value and I was planning on putting it in a context.

Any comments/suggestions about this plan are more than welcome.

File: service/aggregator.go [hunk 4]
```diff
@@ -99,22 +108,30 @@ func (s *AggregateSuggester) GetSuggestions(payload []byte, tid string) (Suggest
...
+	return result
+}
+
 // getSuggestions requests suggestions from the Suggester delegate for the provided payload.
 // It enriches the suggestions with concept data gathered from the ConcordanceService.
 // If the delegate fails to provide suggestions, this function returns suggesterErr error that wraps the delegate error
 // This is done in order to distinguish between errors coming from the Suggester and the ones from ConcordanceService
-func getSuggestions(delegate Suggester, concordance *ConcordanceService, tid string, payload []byte) ([]Suggestion, error) {
-	resp, err := delegate.GetSuggestions(payload, tid)
+func getSuggestions(delegate Suggester, concordance *ConcordanceService, tid, origin string, payload []byte) ([]Suggestion, error) {
```
[the remark this answers is not in the document]
Comment (ivanruski) [reply service/aggregator.go:133]: Is putting them in a context a solution ?

File: service/aggregator.go (hunk 2 above)
Comment (ivanruski) [reply service/aggregator.go]: I've renamed it. (amended the existing commit)

File: service/aggregator.go (hunk 3 above)
Comment (ivanruski) [reply service/aggregator.go]: I've tried to reword it. Please take a look.

File: service/aggregator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivan-p-nikolov) [reply service/aggregator.go:132]: Depends on your world view. If you ask [Dave Cheney](https://dave.cheney.net/2017/01/26/context-is-for-cancelation) contexts should be used only for workflow management.

For me it's complicated. I see the reasoning about tid been in Context, but other things, not so sure. 

As for my suggestion. I was thinking more a long the lines of  
```suggestion
func getSuggestions(delegate Suggester, concordance *ConcordanceService, tid, metadata map[string]string, payload []byte) ([]Suggestion, error) {
```

Where `metadata` is a collection of all `X-` headers.

**Review by ivan-p-nikolov — APPROVED** [review]

**Review by epavlova — APPROVED** [review]

File: reqorigin/reqorigin.go (hunk 1 above)
Comment (vstanchev) [reply reqorigin/reqorigin.go:3]: **thought:** Personally, I avoid future-proofing if it's not going to be immediately beneficial. Here's a [relevant XKCD](https://xkcd.com/974/).
